### PR TITLE
Font sizes are rounded to the nearest 1px

### PR DESCRIPTION
**Pull request by litherum**
#### 30e54c67556479ebd2f316125f8a01e23010d4d9
<pre>
Font sizes are rounded to the nearest 1px
<a href="https://bugs.webkit.org/show_bug.cgi?id=46987">https://bugs.webkit.org/show_bug.cgi?id=46987</a>
&lt;rdar://problem/40829933&gt;

Reviewed by Alan Baradlay.

Previously, we have sibling functions:
float computedSize() const { return m_computedSize; }
unsigned computedPixelSize() const { return unsigned(m_computedSize + 0.5f); }

The bug is that we were using computedPixelSize() rather than computedSize() when performing layout +
rendering, which causes font sizes to be quantized to discrete multiples of 1px. This is problematic
because it means authors can&apos;t get nicely flowing font sizes, as per <a href="https://jsfiddle.net/64hwk8bt/1/.">https://jsfiddle.net/64hwk8bt/1/.</a>
It was also doubly unfortunate because if you did something like divide a font size by 2 and then
multiply it by 2 (in script), the font-size wouldn&apos;t round-trip unperturbed.

This patch fixes this by avoiding the rounding, via migrating callers as such:
+--------------------------------------+------------------------------------------+
| Calls to this:                       | Are replaced with calls to this instead: |
+--------------------------------------+------------------------------------------+
| FontDescription::computedPixelSize() | FontDescription::computedSize()          |
| FontCascade::pixelSize()             | FontCascade::size()                      |
| RenderStyle::computedFontPixelSize() | RenderStyle::computedFontSize()          |
+--------------------------------------+------------------------------------------+

This patch also deletes the old functions which are now unused.

* LayoutTests: Rebaseline test results
* Source/WebCore/accessibility/atspi/AccessibilityObjectTextAtspi.cpp:
(WebCore::AccessibilityObjectAtspi::textAttributes const):
* Source/WebCore/editing/TextIterator.cpp:
(WebCore::shouldEmitExtraNewlineForNode):
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
(WebCore::CanvasRenderingContext2DBase::State::fontString const):
* Source/WebCore/layout/formattingContexts/inline/InlineContentBreaker.cpp:
(WebCore::Layout::InlineContentBreaker::tryBreakingTextRun const):
(WebCore::Layout::InlineContentBreaker::tryHyphenationAcrossOverflowingInlineTextItems const):
* Source/WebCore/layout/formattingContexts/inline/InlineLevelBoxInlines.h:
(WebCore::Layout::m_style):
* Source/WebCore/loader/cache/CachedSVGFont.cpp:
(WebCore::CachedSVGFont::platformDataFromCustomData):
* Source/WebCore/page/PrintContext.cpp:
(WebCore::PrintContext::pageProperty):
* Source/WebCore/page/ios/FrameIOS.mm:
(WebCore::LocalFrame::renderRectForPoint const):
* Source/WebCore/platform/adwaita/ThemeAdwaita.cpp:
(WebCore::ThemeAdwaita::controlSize const):
* Source/WebCore/platform/graphics/FontCascade.h:
(WebCore::FontCascade::pixelSize const): Deleted.
* Source/WebCore/platform/graphics/FontCascadeCache.h:
(WebCore::FontDescriptionKey::FontDescriptionKey):
* Source/WebCore/platform/graphics/FontDescription.cpp:
(WebCore::FontDescription::adjustedSizeForFontFace const):
* Source/WebCore/platform/graphics/FontDescription.h:
(WebCore::FontDescription::computedSize const):
(WebCore::FontDescription::computedPixelSize const): Deleted.
* Source/WebCore/platform/graphics/cocoa/FontCacheCoreText.cpp:
(WebCore::FontCache::createFontPlatformData):
(WebCore::FontCache::lastResortFallbackFont):
* Source/WebCore/platform/graphics/controls/ControlStyle.h:
* Source/WebCore/platform/graphics/coretext/FontCustomPlatformDataCoreText.cpp:
(WebCore::FontCustomPlatformData::fontPlatformData):
* Source/WebCore/platform/graphics/freetype/FontCacheFreeType.cpp:
(WebCore::FontCache::configurePatternForFontDescription):
(WebCore::FontCache::systemFallbackForCharacters):
(WebCore::FontCache::createFontPlatformData):
* Source/WebCore/platform/graphics/freetype/FontCustomPlatformDataFreeType.cpp:
(WebCore::FontCustomPlatformData::fontPlatformData):
* Source/WebCore/platform/graphics/win/FontCacheWin.cpp:
(WebCore::FontCache::lastResortFallbackFont):
(WebCore::FontCache::createFontPlatformData):
* Source/WebCore/platform/graphics/win/FontCustomPlatformDataWin.cpp:
(WebCore::FontCustomPlatformData::fontPlatformData):
* Source/WebCore/platform/mac/ThemeMac.mm:
(WebCore::controlSizeForFont):
(WebCore::stepperControlSizeForFont):
* Source/WebCore/platform/text/Hyphenation.h:
(WebCore::enoughWidthForHyphenation):
* Source/WebCore/platform/win/DragImageWin.cpp:
(WebCore::createDragImageForLink):
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::columnGap const):
* Source/WebCore/rendering/RenderMultiColumnSet.cpp:
(WebCore::RenderMultiColumnSet::columnGap const):
* Source/WebCore/rendering/RenderRubyRun.cpp:
(WebCore::shouldOverhang):
(WebCore::RenderRubyRun::startAndEndOverhang const):
* Source/WebCore/rendering/RenderRubyText.cpp:
(WebCore::RenderRubyText::adjustInlineDirectionLineBounds const):
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::RenderText::maxWordFragmentWidth):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::extractControlStyleForRenderer const):
* Source/WebCore/rendering/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::adjustCheckboxStyle const):
(WebCore::RenderThemeIOS::adjustRadioStyle const):
* Source/WebCore/rendering/RenderThemeMac.mm:
(WebCore::RenderThemeMac::controlSizeForFont const):
(WebCore::RenderThemeMac::controlSizeForSystemFont const):
(WebCore::RenderThemeMac::popupInternalPaddingBox const):
(WebCore::RenderThemeMac::adjustMenuListButtonStyle const):
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::TextBoxPainter&lt;TextBoxPath&gt;::paintBackgroundDecorations):
(WebCore::TextBoxPainter&lt;TextBoxPath&gt;::paintForegroundDecorations):
* Source/WebCore/rendering/line/BreakingContext.h:
(WebCore::tryHyphenating):
* Source/WebCore/rendering/mathml/RenderMathMLRow.cpp:
(WebCore::RenderMathMLRow::stretchVerticalOperatorsAndLayoutChildren):
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::computeLineHeight const):
(WebCore::RenderStyle::computedFontPixelSize const): Deleted.
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/svg/SVGInlineTextBox.cpp:
(WebCore::textShouldBePainted):
* Source/WebCore/rendering/svg/SVGTextLayoutEngineBaseline.cpp:
(WebCore::SVGTextLayoutEngineBaseline::calculateBaselineShift const):
* Source/WebCore/rendering/svg/SVGTextLayoutEngineSpacing.cpp:
(WebCore::SVGTextLayoutEngineSpacing::calculateCSSKerningAndSpacing):
* Source/WebCore/rendering/updating/RenderTreeBuilderFirstLetter.cpp:
(WebCore::styleForFirstLetter):
* Source/WebCore/style/InlineTextBoxStyle.cpp:
(WebCore::computedVisualOverflowForDecorations):
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::resolveElement):
* Source/WebCore/svg/SVGLengthContext.cpp:
(WebCore::SVGLengthContext::convertValueFromUserUnitsToEMS const):
(WebCore::SVGLengthContext::convertValueFromEMSToUserUnits const):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Tools/CISupport/ews-build/steps.py:
(RunWebKitTests):
(RunWebKitTestsInStressMode):
(RunWebKitTestsRedTree):
(RunWebKitTestsWithoutChangeRedTree):

Canonical link: <a href="https://commits.webkit.org/265657@main">https://commits.webkit.org/265657@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/60c5d87abeff58bcac1e8ab76b2ebd1b3ba005a9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11493 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11698 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/12055 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13138 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10958 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/14083 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11679 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13834 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11657 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12524 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/9743 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13560 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9815 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10433 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17587 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10887 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10586 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13776 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10984 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/9049 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10160 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/10312 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2772 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14435 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10841 "Built successfully") | | |

### CODE REVIEW

**Comment by smfr** [comment]
Is this just a rename, or does it fix the bug in the title?

**Comment by litherum** [comment]
It fixes the bug in the title.

```
    float computedSize() const { return m_computedSize; }
    unsigned computedPixelSize() const { return unsigned(m_computedSize + 0.5f); }
```

**Comment by litherum** [comment]
https://jsfiddle.net/64hwk8bt/1/

Screenshot 2022-08-29 at 9 27 11 PM

**Comment by smfr** [comment]
That wasn't clear from the commit message. Can you separate the renames from the bug fix?

**Comment by litherum** [comment]
There are no renames :)

I updated the commit message.

**Comment by litherum** [comment]
I guess I could do better: "This patch migrates:" => "This patch migrates callers from:"

**Comment by litherum** [comment]
Looks like html.css includes

```
font-size: 1.17em;
...
font-size: .83em;
...
font-size: .67em;
```

**Comment by litherum** [comment]
Also we have:

```
// When the CSS keyword "larger" is used, this function will attempt to match within the keyword
// table, and failing that, will simply multiply by 1.2.
inline float BuilderCustom::largerFontSize(float size)
{
    // FIXME: Figure out where we fall in the size ranges (xx-small to xxx-large) and scale up to
    // the next size level.
    return size * 1.2f;
}

// Like the previous function, but for the keyword "smaller".
inline float BuilderCustom::smallerFontSize(float size)
{
    // FIXME: Figure out where we fall in the size ranges (xx-small to xxx-large) and scale down to
    // the next size level.
    return size / 1.2f;
}
```

**Comment by litherum** [comment]
Oh, small caps are also affected because they are 70% of the normal font size.

**Comment by litherum** [comment]
Some tests have `font-size: 250%;` which causes fractional sizes

**Comment by litherum** [comment]
SVG text is also computed to have a fractional font size by default.

**Comment by litherum** [comment]
Superscript and subscript are also a scaled font size.

**Comment by litherum** [comment]
Ruby is scaled too.

**Comment by litherum** [comment]
Looks like this patch is causing SVG text to no longer be drawn :O

File: Source/WebCore/rendering/svg/SVGInlineTextBox.cpp [hunk 1]
```diff
@@ -158,9 +159,7 @@ LayoutRect SVGInlineTextBox::localSelectionRect(unsigned start, unsigned end) co
 
 static inline bool textShouldBePainted(const RenderSVGInlineText& textRenderer)
 {
-    // FontCascade::pixelSize(), returns FontDescription::computedPixelSize(), which returns "int(x + 0.5)".
-    // If the absolute font size on screen is below x=0.5, don't render anything.
-    return textRenderer.scaledFont().pixelSize();
+    return textRenderer.scaledFont().size() < 0.5;
```
Comment (litherum) [inline Source/WebCore/rendering/svg/SVGInlineTextBox.cpp]: Whoopsies

**Comment by graouts** [comment]
> unperterbed

Small spelling mistake here, it's "unperturbed".

Very nice ASCII table.

**Comment by litherum** [comment]
https://github.com/WebKit/WebKit/assets/918903/3abc189e-944b-4914-a367-1d2862d483cd

**Comment by Ahmad-S792** [comment]
@litherum - I think if we import whole SVG directory from WPT as refresh, you wouldn't need to add lots of test case expectations, our SVG directory is outdated and have lot of -manual tests, which don't exist on WPT suite. I have bug for it - [257691](https://bugs.webkit.org/show_bug.cgi?id=257691).

**Comment by litherum** [comment]
The iOS-wk2 failure doesn't look real.

**Comment by smfr** [comment]
"unperterbed" -> "unperturbed"

**Review by alanbaradlay — APPROVED** [review]
while this looks like a great change in general, I am slightly worried about potential repaint artifact fallout, similar to what we saw in the past when going from integral to fractional.

**Comment by litherum** [comment]
> while this looks like a great change in general, I am slightly worried about potential repaint artifact fallout, similar to what we saw in the past when going from integral to fractional.

It's a good point. The state of the world with this patch is:
- Font sizes aren't rounded
- Ascents and descents are still rounded
- Inline-direction math is (and has always been) done on floats

So I think, given the above, this is low enough risk that it's actually okay to land.

**Pull request merged**